File: packages/sdk/src/connectors.ts
/**
 * Project Connectors API — manage credentials at project scope.
 *
 * Maps to `/api/v1/private/projects/:projectId/connectors/*`. Project-scoped
 * connectors are isolated per project, letting separate projects use different
 * API keys for the same service.
 *
 * @example
 * ```ts
 * await melaya.connectors.set("proj_abc", "openai", { value: "sk-..." });
 * const services = await melaya.connectors.connectedServices("proj_abc");
 * ```
 */
import type { HttpClient } from "./client.js";
import type { ConnectedService, ProjectConnectorSetBody } from "./platform-types.js";

export class ConnectorsAPI {
  constructor(private readonly http: HttpClient) {}

  /** List connected services for a project. */
  async connectedServices(projectId: string): Promise<ConnectedService[]> {
    return this.http.get<ConnectedService[]>(
      `/api/v1/private/projects/${encodeURIComponent(projectId)}/connectors/services`,
    );
  }

  /** Store a connector credential at project scope. */
  async set(
    projectId: string,
    service: string,
    body: ProjectConnectorSetBody,
  ): Promise<{ ok: boolean }> {
    return this.http.put<{ ok: boolean }>(
      `/api/v1/private/projects/${encodeURIComponent(projectId)}/connectors/${encodeURIComponent(service)}`,
      body,
    );
  }

  /** Delete a project-scoped connector credential. */
  async delete(projectId: string, service: string): Promise<{ ok: boolean }> {
    return this.http.delete<{ ok: boolean }>(
      `/api/v1/private/projects/${encodeURIComponent(projectId)}/connectors/${encodeURIComponent(service)}`,
    );
  }

  /**
   * Get a short-lived env-handle token for project-scoped credentials.
   * The runner uses this token to decrypt credentials without a full session.
   */
  async envHandle(projectId: string): Promise<{ token: string; expiresAt?: string }> {
    return this.http.post<{ token: string; expiresAt?: string }>(
      `/api/v1/private/projects/${encodeURIComponent(projectId)}/connectors/env-handle`,
    );
  }

  async googleOAuthStart(
    projectId: string,
    body: Record<string, unknown> = {},
  ): Promise<Record<string, unknown>> {
    return this.http.post(
      `/api/v1/private/projects/${encodeURIComponent(projectId)}/connectors/google/oauth`,
      body,
    );
  }
}

File: packages/sdk-rust/src/connectors.rs
use serde_json::{json, Value};
use std::collections::HashMap;

use crate::client::HttpClient;
use crate::error::Result;

/// Project Connectors API — manage credentials at project scope.
#[derive(Clone)]
pub struct ConnectorsAPI {
    http: HttpClient,
}

impl ConnectorsAPI {
    pub(crate) fn new(http: HttpClient) -> Self {
        Self { http }
    }

    /// List connected services for a project.
    pub async fn connected_services(&self, project: &str) -> Result<Value> {
        let q = HashMap::new();
        self.http
            .get(
                &format!("/api/v1/private/projects/{project}/connectors/services"),
                &q,
            )
            .await
    }

    /// Store a connector credential at project scope.
    pub async fn set(
        &self,
        project: &str,
        service: &str,
        value: &str,
        key: Option<&str>,
        label: Option<&str>,
    ) -> Result<Value> {
        let mut body = json!({ "value": value });
        if let Some(k) = key {
            body["key"] = json!(k);
        }
        if let Some(l) = label {
            body["label"] = json!(l);
        }
        self.http
            .put(
                &format!("/api/v1/private/projects/{project}/connectors/{service}"),
                &body,
            )
            .await
    }

    /// Delete a project-scoped connector credential.
    pub async fn delete(&self, project: &str, service: &str) -> Result<Value> {
        let q = HashMap::new();
        self.http
            .delete(
                &format!("/api/v1/private/projects/{project}/connectors/{service}"),
                &q,
            )
            .await
    }

    /// Get a short-lived env-handle token for project-scoped credentials.
    pub async fn env_handle(&self, project: &str) -> Result<Value> {
        self.http
            .post(
                &format!("/api/v1/private/projects/{project}/connectors/env-handle"),
                &json!({}),
            )
            .await
    }

    /// Start Google OAuth flow for project-scoped connector.
    pub async fn google_oauth_start(&self, project: &str, body: &Value) -> Result<Value> {
        self.http
            .post(
                &format!("/api/v1/private/projects/{project}/connectors/google/oauth"),
                body,
            )
            .await
    }
}
